### PR TITLE
Changes to delete RT's if we get an invalid_grant with bad token while renewing tokens

### PR DESCRIPTION
**Pull request by kreedula**

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java [hunk 1]
```diff
@@ -300,6 +302,16 @@ protected void renewAccessToken(@NonNull final SilentTokenCommandParameters para
...
             acquireTokenSilentResult.setLocalAuthenticationResult(authenticationResult);
+        }else {
+            if (tokenResult.getErrorResponse() != null) {
+                final String errorCode = tokenResult.getErrorResponse().getError();
+                final String subErrorCode = tokenResult.getErrorResponse().getSubError();
+
+                if (errorCode.equals(INVALID_GRANT) && subErrorCode.equals(BAD_TOKEN)) {
+                    Logger.info(TAG, "Refresh token is invalid, deleting from cache");
+                    tokenCache.removeCredential(cacheRecord.getRefreshToken());
+                }
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java:313]: Can we log the contents of the `sub_error`?

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java [hunk 2]
```diff
@@ -300,6 +302,16 @@ protected void renewAccessToken(@NonNull final SilentTokenCommandParameters para
 
             // Set the AuthenticationResult on the final result object
             acquireTokenSilentResult.setLocalAuthenticationResult(authenticationResult);
+        }else {
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java:305]: Nit formatter

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java (hunk 1 above)
Comment (kreedula) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java:314]: logging both error and suberror now

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java [hunk 3]
```diff
@@ -300,6 +302,17 @@ protected void renewAccessToken(@NonNull final SilentTokenCommandParameters para
...
+            if (tokenResult.getErrorResponse() != null) {
+                final String errorCode = tokenResult.getErrorResponse().getError();
+                final String subErrorCode = tokenResult.getErrorResponse().getSubError();
+                Logger.info(TAG, "Error: " + errorCode + " Suberror: " + subErrorCode);
+
+                if (errorCode.equals(INVALID_GRANT) && subErrorCode.equals(BAD_TOKEN)) {
+                    Logger.info(TAG, "Refresh token is invalid, deleting from cache");
+                    tokenCache.removeCredential(cacheRecord.getRefreshToken());
+                }
+            }
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java:315]: Can we add one more `else` case here where if the request isn't successful, but _also_ doesn't contain an error response, we log a warning message?

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java:315]: (Obviously, this condition _shouldn't_ ever happen)

Comment (kreedula) [reply common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java:317]: yep, should never happen, but no harm adding

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java [hunk 4]
```diff
@@ -300,6 +302,17 @@ protected void renewAccessToken(@NonNull final SilentTokenCommandParameters para
...
             acquireTokenSilentResult.setLocalAuthenticationResult(authenticationResult);
+        } else {
+            if (tokenResult.getErrorResponse() != null) {
+                final String errorCode = tokenResult.getErrorResponse().getError();
+                final String subErrorCode = tokenResult.getErrorResponse().getSubError();
+                Logger.info(TAG, "Error: " + errorCode + " Suberror: " + subErrorCode);
+
+                if (errorCode.equals(INVALID_GRANT) && subErrorCode.equals(BAD_TOKEN)) {
+                    Logger.info(TAG, "Refresh token is invalid, deleting from cache");
+                    tokenCache.removeCredential(cacheRecord.getRefreshToken());
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java:313]: Let's also check the return value here, should be ` boolean` if this removal was successful

**Pull request merged**